src/haproxy/conn.go

```diff
@@ -0,0 +1,126 @@
+package haproxy
+
+import (
+	"bufio"
+	"net"
+	"time"
+
+	proxyproto "github.com/pires/go-proxyproto"
+)
+
+// Conn is a `net.Conn` compatible struct that handles PROXY header checking.
+type Conn struct {
+	rd  *bufio.Reader
+	c   net.Conn
+	r   net.Addr
+	l   net.Addr
+	hdr *proxyproto.Header
+}
+
+// NewConn wraps a connection with a `bufio.Reader` for checking if the
+// PROXY protocol headers are present
+func NewConn(nc net.Conn) (net.Conn, error) {
+	c := &Conn{
+		c:  nc,
+		rd: bufio.NewReader(nc),
+	}
+	if err := c.ProxyInit(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// ProxyInit checks for the PROXY protocol headers, and if not returns
+// the connection unmolested.
+func (c *Conn) ProxyInit() error {
+	pc, err := proxyproto.Read(c.rd)
+	if err != nil &&
+		err != proxyproto.ErrNoProxyProtocol &&
+		err != proxyproto.ErrInvalidLength {
+		return err
+	} else if err == proxyproto.ErrInvalidLength {
+		// Failed to parse header, the connection is probably about to drop anyway,
+		// it's not a PROXY header at least.
+		return nil
+	}
+
+	c.hdr = pc
+	return nil
+}
+
+// Read reads data from the connection.
+// Read can be made to time out and return an Error with Timeout() == true
+// after a fixed time limit; see SetDeadline and SetReadDeadline.
+func (c *Conn) Read(b []byte) (n int, err error) {
+	return c.rd.Read(b)
+}
+
+// Write writes data to the connection.
+// Write can be made to time out and return an Error with Timeout() == true
+// after a fixed time limit; see SetDeadline and SetWriteDeadline.
+func (c *Conn) Write(b []byte) (n int, err error) {
+	return c.c.Write(b)
+}
+
+// Close closes the connection.
+// Any blocked Read or Write operations will be unblocked and return errors.
+func (c *Conn) Close() error {
+	return c.c.Close()
+}
+
+// LocalAddr returns the local network address.
+func (c *Conn) LocalAddr() net.Addr {
+	if c.hdr == nil {
+		return c.c.LocalAddr()
+	}
+	if c.l == nil {
+		c.l = NewProxyAddr(c.hdr.TransportProtocol, c.hdr.DestinationAddress, c.hdr.DestinationPort)
+	}
+	return c.l
```

Comment:
I'm not 100% sure if it matters for Honeycomb, but the assignment to `c.l` here is potentially racy. (and same for `c.r` below).  I would suggest building these addresses once inside `NewConn()` when a proxy header is present.

Reply:
Thinking with threads, moved the `c.l` and `c.r` assignment into `NewConn()`